### ISSUE TITLE
Analysis: training on all subject data.

### ISSUE THREAD

**Issue by charlesdgburns** [charlesdgburns/NM_TinyRNN#28]
# Overview

After training on all subject data with the same outer fold and inner fold procedure.
**Note**: now the outer fold and inner fold splits will overlap with subject data a lot. This affects some of the results we see below. 

Here we trained models with 
- model_type `GRU` or `monoGRU`, 
- hidden_size `1` or `2` 
- nonlinearities `relu` or `tanh`
- input encodings `unipolar` , `onehot`, or `actonehot`, 
- forced choice being input `_forced` or not ` `.

Our **results** indicate:
1. Models with 2 hidden units fit animal behaviour better than 1 unit models.
2. Greater variability in mechanisms and model dimensionality than when we fit to individual subjects.
3. This variability can be explained by overfitting certain subjects' data leading to lower model dimensionality.

## Fits across models. `2_unit` models fit better than `1_unit` models.
Here showing all models with forced choice being input. Each panel is a combination of nonlinearity and input encoding. The hue is hidden_size, while the x-axis splits results into model type. 

The primary variance in `eval_CE` results from the outer folds, which is similar to the variance seen across subjects when fitting to each subjects' data.

<img width="1583" height="991" alt="Image" src="https://github.com/user-attachments/assets/b36cec41-87a1-4346-b540-5855e081426e" />

## Mechanistic variability across outer folds for `2_unit` model types.
**Note**: Now when we plot mechanistic variability, i.e. the similarity of hidden state activations or gate activations across outer folds, we are technically measuring variability across subjects rather than within each subject (as before).

Here we now see greater variability, surprisingly also for `2_unit_monoGRU_relu_unipolar`, which used to be very consistent when fit to individual subjects' data. In the next section we show that this is due to overfitting to certain subjects that are included in the validation data.

<img width="1991" height="591" alt="Image" src="https://github.com/user-attachments/assets/83c4dfc9-650c-48e2-9790-98cdcf08c50f" />

Taking a closer look at this variability, we see that it is consistent with low dimensionality for some outer x inner fold combinations. For `2_unit_monoGRU_relu_unipolar` we plot the hidden states for each outer fold x inner fold. The hue is `logit_value`. We see several examples of models where the logit is read out along `y=x`, which is a form of dimensionality collapse where the units are highly correlated. 

<img width="3191" height="3441" alt="Image" src="https://github.com/user-attachments/assets/5c3f1c29-1b23-4869-96ed-3022b4fcc476" />

## Low dimensionality in `2_unit_monoGRU_relu_unipolar` stems from validation data overlap with low-d subjects.

**Goal**: we want to measure model dimensionality as a function of which subject contributed to the validation data
**Result**: WS01, WS05, WS09, WS13, and WS20 contribute to 1D model fits when their data takes up a large proportion of the validation dataset.

To check the intuition that mechanistic variability could result from overfitting to specific subjects which take up a large proportion of the validation data, we carried out an analysis on the `2_unit_monoGRU_relu_unipolar` models (N_models=90, with 10 outer folds, x9 inner folds).

First, we measure effective dimensionality (Dim.) as in [Halvagal & Zenke (2023)](https://www.nature.com/articles/s41593-023-01460-y#Sec9). 

```Python
# 1. Filter for validation split
val_trials = trials_data.query('split=="val"')
# 2. Compute eigenvalues of hidden unit activations
Z = val_trials[["hidden_1", "hidden_2"]].to_numpy()
matrix_product = Z.T @ Z
eigenvalues = np.linalg.eigvalsh(matrix_product)
# 3. Compute dimensionality as the participation ratio of each eigenvalue
effective_dim = (np.sum(eigenvalues) ** 2) / np.sum(eigenvalues**2)
```
This gives us a dimensionality score for each model, this will be our explained variable $y$.

Now we loop over each of the 90 models and count the proportion of validation trials that belong to each subject.
For example, outer_fold 1 and inner_fold 1 might have validation data which is 75% from WS02, 25% from WS05, and 0% for all other subjects. This results in a matrix $X$ of size N_subjects X N_models. 

We fit a linear model $y \sim X$, whose coefficients should correspond to the expected dimensionality if 100% of the validation data came from a given subject. We plot the results per subject below (left), with error bars representing 95% confidence intervals in the coefficient. On the right we take a closer look at the data from WS01, seeing how models that are trained with >80% of validation trials from WS01 have effective dimensionality around 1.

<img width="1191" height="491" alt="Image" src="https://github.com/user-attachments/assets/3ca95283-e681-4135-9562-3703d20624cf" />
